File: #0061.rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {

        if (!head || !head->next || k == 0) {
            return head;
        }

        ListNode* ln = head;
        int index = 1;

        while (ln->next) {
            ++index;
            ln = ln->next;
        }

        ln->next = head;
        for (int i = 0; i < index - (k % index); ++i) {
            ln = ln->next;
        }

        head = ln->next;
        ln->next = nullptr;

        return head;
    }
};

File: #0061.rotate-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode RotateRight(ListNode head, int k) {

        if (head == null || head.next == null) {
            return head;
        }

        List<ListNode> ln = new List<ListNode>();
        while (head != null)
        {
            ln.Add(head);
            head = head.next;
        }

        k = k % ln.Count;
        int index = k == 0 ? 0 : ln.Count - k;
        if (index > 0)
        {
            ln[index -1].next = null;
            ln[ln.Count -1].next = ln[0];
        }

        return ln[index];
    }
}
